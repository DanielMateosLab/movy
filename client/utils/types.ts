export interface ShowsByTitleArgs {
  title: string;
  type: string;
  page: number;
}

export type SearchFormValues = Pick<ShowsByTitleArgs, "title" | "type">;
