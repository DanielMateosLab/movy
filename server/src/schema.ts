import { gql } from "apollo-server";

export const typeDefs = gql`
  enum CacheControlScope {
    PUBLIC
    PRIVATE
  }

  directive @cacheControl(
    maxAge: Int
    scope: CacheControlScope
    inheritMaxAge: Boolean
  ) on FIELD_DEFINITION | OBJECT | INTERFACE | UNION

  type Query {
    "Get shows that match a title pattern. We can specify the show type and the results page"
    showsByTitle(title: String!, type: Type, page: Int): ShowsByTitleResponse!
    "Get a single show information by its id"
    showById(id: ID!): Show
  }

  """
  The response of searching a show by title. It contains up to 10 show results,
  the page number, the total results for the query.
  """
  type ShowsByTitleResponse {
    result: [Show]!
    totalResults: Int!
    page: Int!
  }

  "A show is a movie, series or an episode"
  type Show {
    id: ID!
    "The kind of show"
    type: Type!
    title: String!
    "The image url"
    poster: String!
    "The year the show was released"
    year: String!
    "The main theme, like 'terror' or 'fiction'"
    genre: String!
    "A small description or introduction to the show content"
    plot: String!
    "Highlighted actors"
    actors: String!
    "The person in charge of the dramatic and artistic aspects"
    director: String!
    "The person who wrote the scripts"
    writer: String!
    "Highlighted awards"
    awards: String!
  }

  enum Type {
    EPISODE
    SERIES
    MOVIE
  }
`;
