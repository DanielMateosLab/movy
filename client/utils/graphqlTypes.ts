import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export enum CacheControlScope {
  Private = 'PRIVATE',
  Public = 'PUBLIC'
}

export type Query = {
  __typename?: 'Query';
  /** Get a single show information by its id */
  showById?: Maybe<Show>;
  /** Get shows that match a title pattern. We can specify the show type and the results page */
  showsByTitle: ShowsByTitleResponse;
};


export type QueryShowByIdArgs = {
  id: Scalars['ID'];
};


export type QueryShowsByTitleArgs = {
  page?: InputMaybe<Scalars['Int']>;
  title: Scalars['String'];
  type?: InputMaybe<Type>;
};

/** A show is a movie, series or an episode */
export type Show = {
  __typename?: 'Show';
  /** Highlighted actors */
  actors: Scalars['String'];
  /** Highlighted awards */
  awards: Scalars['String'];
  /** The person in charge of the dramatic and artistic aspects */
  director: Scalars['String'];
  /** The main theme, like 'terror' or 'fiction' */
  genre: Scalars['String'];
  id: Scalars['ID'];
  /** A small description or introduction to the show content */
  plot: Scalars['String'];
  /** The image url */
  poster: Scalars['String'];
  title: Scalars['String'];
  /** The kind of show */
  type: Type;
  /** The person who wrote the scripts */
  writer: Scalars['String'];
  /** The year the show was released */
  year: Scalars['String'];
};

/**
 * The response of searching a show by title. It contains up to 10 show results,
 * the page number, the total results for the query.
 */
export type ShowsByTitleResponse = {
  __typename?: 'ShowsByTitleResponse';
  page: Scalars['Int'];
  result: Array<Maybe<Show>>;
  totalResults: Scalars['Int'];
};

export enum Type {
  All = 'ALL',
  Episode = 'EPISODE',
  Movie = 'MOVIE',
  Series = 'SERIES'
}

export type ShowByIdQueryVariables = Exact<{
  id: Scalars['ID'];
}>;


export type ShowByIdQuery = { __typename?: 'Query', showById?: { __typename?: 'Show', poster: string, title: string, type: Type, year: string, plot: string, genre: string, actors: string, director: string, writer: string, awards: string } | null };

export type ShowsByTitleSummaryQueryVariables = Exact<{
  title: Scalars['String'];
  type?: InputMaybe<Type>;
  page?: InputMaybe<Scalars['Int']>;
}>;


export type ShowsByTitleSummaryQuery = { __typename?: 'Query', showsByTitle: { __typename?: 'ShowsByTitleResponse', totalResults: number, page: number, result: Array<{ __typename?: 'Show', id: string, type: Type, title: string, poster: string, year: string } | null> } };


export const ShowByIdDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"ShowById"},"variableDefinitions":[{"kind":"VariableDefinition","variable":{"kind":"Variable","name":{"kind":"Name","value":"id"}},"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"ID"}}}}],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"showById"},"arguments":[{"kind":"Argument","name":{"kind":"Name","value":"id"},"value":{"kind":"Variable","name":{"kind":"Name","value":"id"}}}],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"poster"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"type"}},{"kind":"Field","name":{"kind":"Name","value":"year"}},{"kind":"Field","name":{"kind":"Name","value":"plot"}},{"kind":"Field","name":{"kind":"Name","value":"genre"}},{"kind":"Field","name":{"kind":"Name","value":"actors"}},{"kind":"Field","name":{"kind":"Name","value":"director"}},{"kind":"Field","name":{"kind":"Name","value":"writer"}},{"kind":"Field","name":{"kind":"Name","value":"awards"}}]}}]}}]} as unknown as DocumentNode<ShowByIdQuery, ShowByIdQueryVariables>;
export const ShowsByTitleSummaryDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"ShowsByTitleSummary"},"variableDefinitions":[{"kind":"VariableDefinition","variable":{"kind":"Variable","name":{"kind":"Name","value":"title"}},"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"String"}}}},{"kind":"VariableDefinition","variable":{"kind":"Variable","name":{"kind":"Name","value":"type"}},"type":{"kind":"NamedType","name":{"kind":"Name","value":"Type"}}},{"kind":"VariableDefinition","variable":{"kind":"Variable","name":{"kind":"Name","value":"page"}},"type":{"kind":"NamedType","name":{"kind":"Name","value":"Int"}}}],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"showsByTitle"},"arguments":[{"kind":"Argument","name":{"kind":"Name","value":"title"},"value":{"kind":"Variable","name":{"kind":"Name","value":"title"}}},{"kind":"Argument","name":{"kind":"Name","value":"type"},"value":{"kind":"Variable","name":{"kind":"Name","value":"type"}}},{"kind":"Argument","name":{"kind":"Name","value":"page"},"value":{"kind":"Variable","name":{"kind":"Name","value":"page"}}}],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"result"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"type"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"poster"}},{"kind":"Field","name":{"kind":"Name","value":"year"}}]}},{"kind":"Field","name":{"kind":"Name","value":"totalResults"}},{"kind":"Field","name":{"kind":"Name","value":"page"}}]}}]}}]} as unknown as DocumentNode<ShowsByTitleSummaryQuery, ShowsByTitleSummaryQueryVariables>;