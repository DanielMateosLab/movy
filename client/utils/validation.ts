import * as yup from "yup";

export const searchShowByTitleValidation = yup.object({
  title: yup.string().required("A title is required"),
  type: yup.string(),
  page: yup.number().min(1).integer(),
});
