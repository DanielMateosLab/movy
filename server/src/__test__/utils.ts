export const toReadableJSON = (value: any) =>
  JSON.stringify(value, undefined, 2);
