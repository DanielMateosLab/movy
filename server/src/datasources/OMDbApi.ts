import { RESTDataSource } from "apollo-datasource-rest";
import { UserInputError } from "apollo-server";
import { OMDb_API_KEY } from "../config";
import {
  ApiShow,
  GetShowByIdApiResponse,
  SearchShowsByTitleApiResponse,
} from "../types";
import { QueryShowByIdArgs, QueryShowsByTitleArgs } from "../types/graphql";

class OMDbApi extends RESTDataSource {
  constructor() {
    super();

    this.baseURL = "http://www.omdbapi.com/";
  }

  async searchShowsByTitle({ title, type, page }: QueryShowsByTitleArgs) {
    const result: SearchShowsByTitleApiResponse = await this.get("", {
      s: title,
      type: type || "",
      page: page || 1,
      apikey: OMDb_API_KEY,
    });

    if (result.Response == "True") {
      return {
        result: result.Search,
        totalResults: result.totalResults,
        page: page || 1,
      };
    }

    throw new UserInputError(result.Error);
  }

  async getShowById({ id }: QueryShowByIdArgs): Promise<ApiShow | null> {
    const result: GetShowByIdApiResponse = await this.get("", {
      i: id,
      apiKey: OMDb_API_KEY,
    });

    if (result.Response == "True") {
      const { Response, ...show } = result;

      return show;
    }

    return null;
  }

  /**
   * Returns parent[property] if it is defined.
   * If not, calls OMDbApi "search by id" endpoint to get "property"
   * and returns it.
   */
  async getPropertyFromParentOrAPI(property: keyof ApiShow, parent: ApiShow) {
    if (property in parent) return property;

    return (await this.getShowById({ id: parent.imdbID }))![property];
  }
}

export default OMDbApi;
