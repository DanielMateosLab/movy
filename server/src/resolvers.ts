import { ApiShow, AppContextType, ShowType } from "./types";
import { Resolvers } from "./types/graphql";

const resolvers: Resolvers = {
  Query: {
    showsByTitle: (_, { title, type, page }, { dataSources }) => {
      return dataSources.omdbApi.searchShowsByTitle({ title, type, page });
    },
  },
  Show: {
    id: ({ imdbID }) => imdbID,
    type: ({ Type }) => Type,
    title: ({ Title }) => Title,
    poster: ({ Poster }) => Poster,
    year: ({ Year }) => Year,
    ...generateParentOrAPIResolverForProperties(
      "Genre",
      "Plot",
      "Director",
      "Actors",
      "Writer",
      "Awards"
    ),
  },
  Type: ShowType,
};

/** Helper function that generates, for each property given, a resolver that will:
 *
 * - Try to return the property from the parents arg.
 * - Call OMDbAPI rest API searching the show by id, and return the asked
 *     show property.
 *
 * Those resolvers will be returned in an object whose keys are the properties
 * in lower case, so they match our Show schema type.
 */
function generateParentOrAPIResolverForProperties(
  ...properties: Array<keyof ApiShow>
) {
  return properties.reduce(
    (result, property) => ({
      ...result,
      // We return the property as lowercase to match our schema "Show"
      [String(property).toLowerCase()]: generateParentOrAPIResolver(property),
    }),
    {}
  );
}

/** For the given property, generates a resolver that uses OMDbAPI.getPropertyFromParentOrAPI() */
function generateParentOrAPIResolver(property: keyof ApiShow) {
  return (parent: ApiShow, _: undefined, { dataSources }: AppContextType) => {
    return dataSources.omdbApi.getPropertyFromParentOrAPI(property, parent);
  };
}

export default resolvers;
