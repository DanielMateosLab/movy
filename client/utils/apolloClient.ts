import { ApolloClient, InMemoryCache, makeVar } from "@apollo/client";
import { Type } from "./graphqlTypes";
import { ShowsByTitleArgs } from "./types";

export const showsByTitleArgs = makeVar<ShowsByTitleArgs>({
  title: "",
  type: Type.All,
  page: 1,
});

const apolloClient = new ApolloClient({
  uri: process.env.NEXT_PUBLIC_SERVER_URI,
  cache: new InMemoryCache(),
});

export default apolloClient;
