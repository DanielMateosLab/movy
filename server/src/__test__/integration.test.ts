import { ApolloServer } from "apollo-server";
import resolvers from "../resolvers";
import { typeDefs } from "../schema";
import OMDbApi from "../datasources/OMDbApi";
import {
  errorMockResponse,
  showByIdMockResponse,
  showsByTitleMockRes,
} from "./mockApiResults";
import { toReadableJSON } from "./utils";

/** Here resolvers and dataSources are tested in conjunction.
 *  To do so, I mock the response from the get() method of the RESTDataSource
 *  class with real responses of the OMDbAPI.
 */

describe("Queries", () => {
  let omdbApi: OMDbApi;
  let testServer: ApolloServer;

  beforeEach(() => {
    omdbApi = new OMDbApi();

    testServer = new ApolloServer({
      typeDefs,
      resolvers,
      dataSources: () => ({ omdbApi }),
    });
  });
  describe("Query.showsByTitle", () => {
    describe("Simple Search Show Query", () => {
      // This query tries to fetch only for the fields we get from a normal
      // "search by title" OMDbAPI request.
      const operation = {
        query: `
        query TestQuery($title: String!) {
          showsByTitle(title: $title) {
            result {
              title
              type
              id
              poster
              year
            }
          totalResults
          page
            }
          }
        `,
        variables: { title: "mock" },
      };

      it("gets a list of shows", async () => {
        // Mock a success response from API
        // @ts-ignore - We skip checks here because get is a protected method
        // and we can not call .mockImplementationValueOnce() on it
        OMDbApi.prototype.get = jest.fn(async () => showsByTitleMockRes);

        const result = await testServer.executeOperation(operation);

        expect(result.errors).toBeUndefined();
        expect(result.data).toBeDefined();
        expect(result).toMatchSnapshot();
      });

      it("returns an error if with an invalid search string", async () => {
        // @ts-ignore
        OMDbApi.prototype.get = jest.fn(async () => errorMockResponse);

        const result = await testServer.executeOperation(operation);

        expect(result.data).toBeNull();
        expect(result.errors).toBeDefined();
        expect(result).toMatchSnapshot();
      });
    });

    describe("Complex Search Query", () => {
      // This query tries to fetch all the details of a show, so additional
      // requests to "show by id" OMDbAPI endpoints are needed.
      const operation = {
        query: `
         query TestQuery($title: String!) {
            showsByTitle(title: $title) {
              result {
                id
                type
                title
                poster
                year
                genre
                plot
                director
                actors
                writer
                awards
              }
            }
          }
        `,
        variables: { title: "mock" },
      };

      it("gets a list of shows", async () => {
        // We mock .get() method so it differenciates between searchs by title or by id
        const mockGet = jest.fn(async (_, queryParams) => {
          // If "i" is in queryParams, we are searching by id
          if ("i" in queryParams) return showByIdMockResponse;
          // Else, we are searching by title
          return showsByTitleMockRes;
        });

        // @ts-ignore
        OMDbApi.prototype.get = mockGet;

        const result = await testServer.executeOperation(operation);

        console.log(toReadableJSON(result));

        expect(result.errors).toBeUndefined();
        expect(result.data).toBeDefined();
      });
    });
  });
  describe.skip("Query.showById", () => {});
});
