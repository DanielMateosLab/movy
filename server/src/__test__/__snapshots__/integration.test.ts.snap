// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries Query.showsByTitle Simple Search Show Query gets a list of shows 1`] = `
Object {
  "data": Object {
    "showsByTitle": Object {
      "page": 1,
      "result": Array [
        Object {
          "id": "tt0372784",
          "poster": "https://m.media-amazon.com/images/M/MV5BOTY4YjI2N2MtYmFlMC00ZjcyLTg3YjEtMDQyM2ZjYzQ5YWFkXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg",
          "title": "Batman Begins",
          "type": "MOVIE",
          "year": "2005",
        },
        Object {
          "id": "tt2975590",
          "poster": "https://m.media-amazon.com/images/M/MV5BYThjYzcyYzItNTVjNy00NDk0LTgwMWQtYjMwNmNlNWJhMzMyXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg",
          "title": "Batman v Superman: Dawn of Justice",
          "type": "MOVIE",
          "year": "2016",
        },
        Object {
          "id": "tt0096895",
          "poster": "https://m.media-amazon.com/images/M/MV5BMTYwNjAyODIyMF5BMl5BanBnXkFtZTYwNDMwMDk2._V1_SX300.jpg",
          "title": "Batman",
          "type": "MOVIE",
          "year": "1989",
        },
        Object {
          "id": "tt0103776",
          "poster": "https://m.media-amazon.com/images/M/MV5BOGZmYzVkMmItM2NiOS00MDI3LWI4ZWQtMTg0YWZkODRkMmViXkEyXkFqcGdeQXVyODY0NzcxNw@@._V1_SX300.jpg",
          "title": "Batman Returns",
          "type": "MOVIE",
          "year": "1992",
        },
        Object {
          "id": "tt0112462",
          "poster": "https://m.media-amazon.com/images/M/MV5BNDdjYmFiYWEtYzBhZS00YTZkLWFlODgtY2I5MDE0NzZmMDljXkEyXkFqcGdeQXVyMTMxODk2OTU@._V1_SX300.jpg",
          "title": "Batman Forever",
          "type": "MOVIE",
          "year": "1995",
        },
        Object {
          "id": "tt0118688",
          "poster": "https://m.media-amazon.com/images/M/MV5BMGQ5YTM1NmMtYmIxYy00N2VmLWJhZTYtN2EwYTY3MWFhOTczXkEyXkFqcGdeQXVyNTA2NTI0MTY@._V1_SX300.jpg",
          "title": "Batman & Robin",
          "type": "MOVIE",
          "year": "1997",
        },
        Object {
          "id": "tt1877830",
          "poster": "https://m.media-amazon.com/images/M/MV5BMDdmMTBiNTYtMDIzNi00NGVlLWIzMDYtZTk3MTQ3NGQxZGEwXkEyXkFqcGdeQXVyMzMwOTU5MDk@._V1_SX300.jpg",
          "title": "The Batman",
          "type": "MOVIE",
          "year": "2022",
        },
        Object {
          "id": "tt4116284",
          "poster": "https://m.media-amazon.com/images/M/MV5BMTcyNTEyOTY0M15BMl5BanBnXkFtZTgwOTAyNzU3MDI@._V1_SX300.jpg",
          "title": "The Lego Batman Movie",
          "type": "MOVIE",
          "year": "2017",
        },
        Object {
          "id": "tt1569923",
          "poster": "https://m.media-amazon.com/images/M/MV5BNmY4ZDZjY2UtOWFiYy00MjhjLThmMjctOTQ2NjYxZGRjYmNlL2ltYWdlL2ltYWdlXkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
          "title": "Batman: Under the Red Hood",
          "type": "MOVIE",
          "year": "2010",
        },
        Object {
          "id": "tt4853102",
          "poster": "https://m.media-amazon.com/images/M/MV5BMTdjZTliODYtNWExMi00NjQ1LWIzN2MtN2Q5NTg5NTk3NzliL2ltYWdlXkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
          "title": "Batman: The Killing Joke",
          "type": "MOVIE",
          "year": "2016",
        },
      ],
      "totalResults": 406,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries Query.showsByTitle Simple Search Show Query returns an error if with an invalid search string 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Movie not found!],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
